File: java/sorts/quickSort/QuickSort2.java
import java.util.List;

public class QuickSort2 {

    public static void main(String[] args) {
        List<Integer> numbers = FileUtil.readNumbersFromFile("./input_sort.txt");
        Integer[] arr = numbers.toArray(new Integer[0]);

        long start = System.nanoTime();
        quickSort(arr, 0, arr.length - 1);
        long end = System.nanoTime();

        System.out.println("MS : " + (double) (end - start) / 1_000_000);
        
        FileUtil.writeArrayToFile("./output_quick_sort.txt", arr);
    }

    private static void quickSort(Integer[] A, int p, int r) {
        if (p < r) {
            int q = partition(A, p, r);
            quickSort(A, p, q - 1);
            quickSort(A, q + 1, r);
        }
    }

    private static int partition(Integer[] A, int p, int r) {
        final int mid = (p + r) >>> 1;
        final int x = A[mid];
        int i = p - 1;
        for (int j = p; j <= r - 1; j++) {
            if (A[j] <= x) {
                i++;
                swap(A, i, j);
            }
        }
        swap(A, i + 1, r); // 피벗을 제자리에 놓기
        return i + 1;
    }

    private static void swap(Integer[] A, int i, int j) {
        int temp = A[i];
        A[i] = A[j];
        A[j] = temp;
    }
}

File: c/sorts/quickSort/QuickSort2.c
#include <time.h>
#include <stdio.h>

void swap(int* a, int* b) {
    int temp;
    temp = *a;
    *a = *b;
    *b = temp;
}

int partition(int A[], int l, int r) {
    int pivot, i, j;
    pivot = A[r];
    i = l - 1;
    for(j = l; j <= r-1; j++) {
        if(A[j] <= pivot) {
            i++;
            swap(&A[i], &A[j]);
        }
    }
    swap(&A[i+1], &A[r]);
    return i+1;
}

void quickSort(int A[], int l, int r) {
    if (l < r) {
        int pivotIdx = partition(A, l, r);
        quickSort(A, l, pivotIdx-1);
        quickSort(A, pivotIdx+1, r);
    }
}

int main() {
    clock_t start, end;
    int array[] = {2, 5, 7, 6, 1, 8, 9, 3, 4, 10};
    int size = sizeof(array) / sizeof(array[0]);

    start = clock();
    quickSort(array, 0, size-1);
    end = clock();

    // 정렬된 배열 출력
    printf("Sorted array: ");
    for (int i = 0; i < size; i++) {
        printf("%d ", array[i]);
    }
    printf("\n");

    printf("MS : %lf\n", (double)(end - start)/CLOCKS_PER_SEC*1000);
    
    return 0;
}